design/restore-with-EnableAPIGroupVersions-feature.md

```diff
@@ -0,0 +1,135 @@
+# Restore API Group Version by Priority Level When EnableAPIGroupVersions Feature is Set
+
+Status: Draft
+
+## Abstract
+
+This document proposes a solution to select an API group version to restore from the versions backed up using the feature EnableAPIGroupVersions.
+
+## Background
+
+It is possible that between the time a backup has been made and a restore occurs that the target Kubernetes version has incremented more than one version. In such a case where multiple versions of Kubernetes were skipped, the preferred source cluster's API group versions for resources may no longer be supported by the target cluster. With PR #2373, all supported API group versions were backed up if the EnableAPIGroupVersions feature flag was set during Velero installation. The next step (outlined by this design proposal) will be to see if any of the backed up versions are supported in the target cluster and if so, choose one to restore for each backed up resource.
+
+## Goals
+
+- Restore resources using the chosen API group version.
+
+## Non Goals
+
+- Allowing users to customize the predetermined API group versions priorities.
+- Modifying the compressed backup tarball files.
+- Using plugins to restore a resource when the target supports non of the source cluster's API group versions.
+
+## High-Level Design
+
+During restore, the proposal is that Velero will read the directory names in the backup tarball file and see which API group versions were backed up from the source cluster. Velero will also look at the target cluster and determine which API group versions are supported. Given the two lists of source and target cluster supported API versions, a priority system will be used to determine which version will be backed up. The priority of versions in order are target cluster preferred version, source cluster preferred version, and the latest common supported version. Once the version to restore is chosen, the file path to the backed up resource in the tarball will be modified such that it points to the resources' chosen API group version. If no version is found in common between the source and target clusters, the chosen version will default to the source cluster's preferred version (the version being restored currently without the changes proposal here). Restore will be allowed to continue as before.
+
+## Detailed Design
+
+There are four main objectives to achieve the above stated goal:
+
+1. List the backed up API group versions.
+1. List the API group versions supported by the target cluster.
+1. Use a priority system to determine which version to back up with the source preferred version as the default.
```

Comment:
I was thought that step 3 would be `Use a priority system to determine which version to restore from with the target preferred version as the default.`

Because the priority system will determine which API version to be restored from by:
1. getting source API versions from the backup tarball
2. getting target API versions by discovery API
3. determine which API version to be restored from the backup tarball with order
   - target cluster preferred version if found in the backup tarball
   - source cluster preferred version if source cluster preferred version available in the target cluster
   - the latest common supported version for others

Reply:
The default will only take effect if all the three priorities fail. I chose `source preferred` so we can revert back to what the logic was like before the code changes with this PR. When our priority system fails, the failure will look the same as it would have failed before the feature flag was enabled.

Here are the sequence of events in case of failure:

1. Set default to `source cluster preferred version`.
2. Priority one fails. We cannot find a target preferred version that is supported by the source versions.
3. Priority two fails. We cannot find a source preferred version that is supported by the target versions.
4. Priority three fails. We cannot find a version that is common to the target and source.
5. Restore will use default and continue to report error per normal (backup version not supported by target cluster or something like that).

Reply:
> The default will only take effect if all the three priorities fail. I chose `source preferred` so we can revert back to what the logic was like before the code changes with this PR. When our priority system fails, the failure will look the same as it would have failed before the feature flag was enabled.
> 
> Here are the sequence of events in case of failure:
> 
> 1. Set default to `source cluster preferred version`.
> 2. Priority one fails. We cannot find a target preferred version that is supported by the source versions.
> 3. Priority two fails. We cannot find a source preferred version that is supported by the target versions.
> 4. Priority three fails. We cannot find a version that is common to the target and source.
> 5. Restore will use default and continue to report error per normal (backup version not supported by target cluster or something like that).

Thanks for your clear explanation, a good idea if all the three priorities fail, it falls back to use `source cluster preferred version`.
I do not understand the sentence at my first glance `Use a priority system to determine which version to back up with the source preferred version as the default.`

Reply:
I admit that sentence is confusing. I'll split it up. Is this easier to understand:

`Use a priority system to determine which version to back up. The source preferred version will be the default if the priorities fail.`?

Reply:
Use a priority system to determine which version to "back up", should it be "restore from" 🤔 ?

Reply:
haha oops. Another good catch @jenting. Should be 
> Use a priority system to determine which version to restore with the source preferred version as the default.